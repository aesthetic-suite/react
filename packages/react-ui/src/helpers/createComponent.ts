/* eslint-disable @typescript-eslint/no-explicit-any */

import { forwardRef } from 'react';
import { CommonHtmlProps, HtmlElementType, OmitUnwantedHtmlProps } from '../types';

export function createComponent<Props extends object, Element extends HTMLElement = HTMLElement>(
	component: React.ForwardRefRenderFunction<Element, Props>,
) {
	return forwardRef(component);
}

export type DynamicProps<
	As extends React.ElementType,
	Props extends object
> = OmitUnwantedHtmlProps<React.ComponentPropsWithRef<As>> &
	Props & {
		/**
		 * Native HTML element or custom component to render as.
		 */
		as?: As;
	};

// This is based on the `React.ExoticComponent` but supports the dynamic `as` prop
// through a generic on the render function. We want the props on the consumer to change
// based on the `as` prop, but we *do not* want to type the props/ref in the implementation,
// as it makes the types too complex and we spread them into the DOM regardless.
export interface DynamicComponent<Props extends object, As extends React.ElementType> {
	<T extends As>(props: DynamicProps<T, Props>): React.ReactElement | null;
	readonly $$typeof: symbol;
	displayName?: string;
}

// We only want to return generic attributes when an HTML element,
// otherwise return the props from the custom component.
export type InferAsProps<T> = T extends HtmlElementType
	? CommonHtmlProps
	: T extends React.JSXElementConstructor<infer P>
	? P
	: {};

export function createDynamicComponent<
	Props extends object,
	As extends React.ElementType = HtmlElementType
>(component: React.ForwardRefRenderFunction<any, InferAsProps<As> & Props & { as?: As }>) {
	return (createComponent(component) as unknown) as DynamicComponent<Props, As>;
}
