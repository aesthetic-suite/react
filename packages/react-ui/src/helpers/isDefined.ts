export function isDefined(...values: unknown[]): boolean {
	return values.some((value) => typeof value !== 'undefined' && value !== null && value !== false);
}
