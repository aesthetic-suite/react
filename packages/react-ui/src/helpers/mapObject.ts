import { objectLoop } from '@aesthetic/utils';

export function mapObject<P, N>(
  base: Record<string, P>,
  onValue: (value: P) => N | undefined,
  onKey?: (key: string) => string,
): Record<string, N> {
  const object: Record<string, unknown> = {};

  objectLoop(base, (value, key) => {
    const nextValue = onValue(value);

    if (nextValue !== undefined) {
      object[onKey ? onKey(key) : key] = nextValue;
    }
  });

  return object as Record<string, N>;
}
