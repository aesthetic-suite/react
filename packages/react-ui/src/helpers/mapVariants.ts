import { ElementStyles } from '@aesthetic/react';
import { arrayLoop } from '@aesthetic/utils';

export function mapVariants<T>(
	type: string,
	enums: T[],
	factory: (value: T, typeName: string) => ElementStyles,
) {
	const object: Partial<Record<string, ElementStyles>> = {};

	arrayLoop(enums, (value) => {
		object[`${type}:${value}`] = factory(value, type);
	});

	return object as Record<string, ElementStyles>;
}
