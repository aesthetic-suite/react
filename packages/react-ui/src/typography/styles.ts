import { PALETTE_TYPES, styles } from '@aesthetic/react';
import { mapVariants } from '../helpers/mapVariants';

export const styleSheet = styles((css) => {
	function weight(size: number | 'inherit') {
		return {
			fontWeight: size,
		};
	}

	return {
		typography: {
			color: css.var('palette-neutral-text'),
			margin: 0,
			padding: 0,

			'@variants': {
				// Alignment
				'align:start': {
					textAlign: 'start',
					'@fallbacks': {
						textAlign: 'left',
					},
				},
				'align:center': {
					textAlign: 'center',
				},
				'align:end': {
					textAlign: 'end',
					'@fallbacks': {
						textAlign: 'right',
					},
				},
				'align:justify': {
					textAlign: 'justify',
				},

				// Overflow
				'overflow:break': css.mixin('text-break'),
				'overflow:clip': css.mixin('text-truncate', {
					textOverflow: 'clip',
				}),
				'overflow:truncate': css.mixin('text-truncate'),
				'overflow:wrap': css.mixin('text-wrap'),

				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
				})),

				// Transforms
				'transform:capitalize': { textTransform: 'capitalize' },
				'transform:lowercase': { textTransform: 'lowercase' },
				'transform:uppercase': { textTransform: 'uppercase' },

				// Font weights
				'weight:inherit': weight('inherit'),
				'weight:thin': weight(100),
				'weight:light': weight(300),
				'weight:normal': weight(400),
				'weight:bold': weight(600),
				'weight:black': weight(800),
			},
		},

		monospace: {
			fontFamily: css.var('typography-font-monospace'),
		},

		link: {
			display: 'inline-block',
			backgroundColor: 'transparent',
			borderColor: 'transparent',
			borderWidth: 0,
			cursor: 'pointer',
			textDecoration: 'none',
			fontFamily: 'inherit',
			fontSize: 'inherit',
			lineHeight: 'inherit',

			'@variants': {
				// Palettes
				...mapVariants('palette', PALETTE_TYPES, (palette) => ({
					color: css.var(`palette-${palette}-text`),
					'@selectors': {
						':visited': { color: css.var(`palette-${palette}-color-60`) },
						':hover': { color: css.var(`palette-${palette}-color-80`) },
						':active': { color: css.var(`palette-${palette}-color-90`) },
					},
				})),
			},
		},
	};
});
