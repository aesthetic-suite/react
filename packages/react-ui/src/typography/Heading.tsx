import React, { useMemo } from 'react';
import { HeadingLevel, useStyles } from '@aesthetic/react';
import { createDynamicComponent } from '../helpers/createComponent';
import { useHeading } from '../hooks/styles/useHeading';
import { CommonProps } from '../types';
import { styleSheet } from './styles';
import { TypographyProps } from './types';
import { TypographyContext } from './TypographyContext';

export * from './types';

export type HeadingElement = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export interface HeadingProps extends TypographyProps, CommonProps {
	/**
	 * Increase or decrease the font size, line height, and letter spacing.
	 */
	level: HeadingLevel;
}

/**
 * @link https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#heading_content
 */
export const Heading = createDynamicComponent<HeadingProps, HeadingElement>(function Heading(
	{
		align = 'start',
		as,
		children,
		overflow = 'wrap',
		palette = 'neutral',
		level = 1,
		testID,
		transform,
		weight = 'normal',
		// Inherited
		className: inheritedClassName,
		...props
	},
	ref,
) {
	const cx = useStyles(styleSheet);
	const headingClassName = useHeading(level);
	const className = useMemo(
		() =>
			cx({ align, overflow, palette, transform, weight }, 'typography', [
				headingClassName,
				inheritedClassName,
			]),
		[cx, align, overflow, palette, transform, weight, headingClassName, inheritedClassName],
	);
	const Tag = as ?? `h${level}`!;

	return (
		<Tag ref={ref} data-testid={testID} {...props} className={className}>
			<TypographyContext.Provider value>{children}</TypographyContext.Provider>
		</Tag>
	);
});
