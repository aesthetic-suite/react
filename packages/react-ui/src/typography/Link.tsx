import React, { useMemo } from 'react';
import { useStyles } from '@aesthetic/react';
import { createComponent } from '../helpers/createComponent';
import { useContextRequirement } from '../hooks/useContextRequirement';
import { HtmlAnchorProps, Pressable } from '../internal/Pressable';
import { CommonProps } from '../types';
import { styleSheet } from './styles';
import { TypographyProps } from './types';
import { TypographyContext } from './TypographyContext';

export * from './types';

export interface LinkProps
	extends Pick<TypographyProps, 'children' | 'palette' | 'transform' | 'weight'>,
		Omit<HtmlAnchorProps, 'children' | 'href' | 'onClick'>,
		CommonProps {
	/**
	 * Document or address to link to.
	 */
	to?: string;
	/**
	 * Callback fired when the link is clicked.
	 */
	onClick?: React.MouseEventHandler<HTMLElement>;
}

export const Link = createComponent<LinkProps>(function Link(
	{
		children,
		palette = 'primary',
		transform,
		weight = 'normal',
		// Inherited
		className: inheritedClassName,
		...props
	},
	ref,
) {
	const cx = useStyles(styleSheet);
	const className = useMemo(
		() => cx({ palette, transform, weight }, 'typography', 'link', [inheritedClassName]),
		[cx, palette, transform, weight, inheritedClassName],
	);

	useContextRequirement(
		TypographyContext,
		__DEV__ ? '`Link` component must be rendered within a `Text` or `Heading` component.' : '',
	);

	return (
		<Pressable ref={ref} {...props} className={className}>
			{children}
		</Pressable>
	);
});
