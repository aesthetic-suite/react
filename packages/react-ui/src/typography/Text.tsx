import React, { useMemo } from 'react';
import { TextSize, useStyles } from '@aesthetic/react';
import { createDynamicComponent } from '../helpers/createComponent';
import { useText } from '../hooks/styles/useText';
import { CommonProps } from '../types';
import { styleSheet } from './styles';
import { TypographyProps } from './types';

export * from './types';

export type TextElement =
	| 'abbr'
	| 'b'
	| 'bdo'
	| 'cite'
	| 'code'
	| 'data'
	| 'dfn'
	| 'div'
	| 'em'
	| 'i'
	| 'kbd'
	| 'p'
	| 'q'
	| 'samp'
	| 'small'
	| 'span'
	| 'strong'
	| 'sub'
	| 'sup'
	| 'time'
	| 'var'
	| 'wbr';

export interface TextProps extends TypographyProps, CommonProps {
	/**
	 * Switch to a monospace font family defined in the current design system theme.
	 */
	monospaced?: boolean;
	/**
	 * Increase or decrease the font size and line height.
	 * @default df
	 */
	size?: TextSize;
}

/**
 * @link https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content
 */
export const Text = createDynamicComponent<TextProps, TextElement>(function Text(
	{
		align = 'start',
		as: Tag = 'p',
		children,
		monospaced,
		overflow = 'wrap',
		palette = 'neutral',
		size = 'df',
		testID,
		transform,
		weight = 'normal',
		// Inherited
		className: inheritedClassName,
		...props
	},
	ref,
) {
	const cx = useStyles(styleSheet);
	const textClassName = useText(size);
	const className = useMemo(
		() =>
			cx(
				{ align, overflow, palette, size, transform, weight },
				'typography',
				monospaced && 'monospace',
				[textClassName, inheritedClassName],
			),
		[
			cx,
			align,
			overflow,
			palette,
			size,
			transform,
			weight,
			monospaced,
			textClassName,
			inheritedClassName,
		],
	);

	return (
		<Tag ref={ref} data-testid={testID} {...props} className={className}>
			{children}
		</Tag>
	);
});
