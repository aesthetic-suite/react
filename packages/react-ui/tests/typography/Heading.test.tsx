import React from 'react';
import { render, screen } from '@testing-library/react';
import { Heading, HeadingProps } from '../../src/typography/Heading';
import { getRenderedStyles, withAccessibility, withStyles, Wrapper } from '../helpers';

function getElement() {
	return screen.getByText('Title')!;
}

describe('Heading', () => {
	it('renders the title', () => {
		render(<Heading level={1}>Title</Heading>, { wrapper: Wrapper });

		expect(getElement().className).toBe(
			'typography variant:align:start variant:overflow:wrap variant:palette:neutral variant:weight:normal variant:level:1',
		);
	});

	it(
		'renders the title styles',
		withStyles(() => {
			render(<Heading level={1}>Title</Heading>, { wrapper: Wrapper });

			expect(getRenderedStyles('standard')).toMatchSnapshot();
		}),
	);

	it(
		'passes accessibility',
		withAccessibility(async (axe) => {
			const { container } = render(<Heading level={1}>Title</Heading>, { wrapper: Wrapper });

			expect(await axe(container)).toHaveNoViolations();
		}),
	);

	([1, 2, 3, 4, 5, 6] as const).forEach((level) => {
		it(`renders level ${level}`, () => {
			render(<Heading level={level}>Title</Heading>, { wrapper: Wrapper });

			expect(getElement().tagName).toBe(`H${level}`);
		});
	});

	it('can change element outside of level', () => {
		render(
			<Heading as="h4" level={1}>
				Title
			</Heading>,
			{ wrapper: Wrapper },
		);

		expect(getElement().tagName).toBe('H4');
	});

	it('can pass a test ID', () => {
		render(
			<Heading level={1} testID="test">
				Title
			</Heading>,
			{ wrapper: Wrapper },
		);

		expect(getElement().getAttribute('data-testid')).toBe('test');
	});

	it('can pass a custom class name', () => {
		render(
			<Heading className="foobar" level={1}>
				Title
			</Heading>,
			{ wrapper: Wrapper },
		);

		expect(getElement().className).toEqual(expect.stringContaining('foobar'));
	});

	it('can pass native attributes', () => {
		render(
			<Heading aria-label="Label" id="foo" level={1}>
				Title
			</Heading>,
			{ wrapper: Wrapper },
		);

		const el = getElement();

		expect(el.id).toBe('foo');
		expect(el.getAttribute('aria-label')).toBe('Label');
	});

	describe('props', () => {
		const props: Partial<HeadingProps> = {
			align: 'center',
			level: 3,
			overflow: 'break',
			palette: 'danger',
			transform: 'capitalize',
			weight: 'bold',
		};

		Object.entries(props).forEach(([prop, value]) => {
			it(`sets "${prop}" class name`, () => {
				render(
					<Heading level={1} {...{ [prop]: value }}>
						Title
					</Heading>,
					{ wrapper: Wrapper },
				);

				expect(getElement().className).toEqual(
					expect.stringContaining(`variant:${prop}:${value as string}`),
				);
			});
		});
	});
});
