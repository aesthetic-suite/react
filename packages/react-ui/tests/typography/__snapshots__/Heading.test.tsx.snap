// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders the title with default styles 1`] = `".c1u0bpqh {color: var(--palette-neutral-text);}.c13kbekr {margin: 0;}.cqmlbj8 {padding: 0;}.c1p4ubeg {text-align: start;}.cngw5jn {text-align: center;}.cix8cjb {text-align: end;}.caz11cm {text-align: justify;}.cnvui58 {overflow-wrap: break-word;}.cc50pji {word-wrap: break-word;}.c1gvf4w5 {word-break: break-word;}.cs4wsyb {overflow: hidden;}.csbu2xr {text-overflow: clip;}.c1a0rdy6 {white-space: nowrap;}.c1804rdo {text-overflow: ellipsis;}.c1xww3tf {overflow-wrap: normal;}.c1f1iw5d {word-wrap: normal;}.cn25lgq {word-break: normal;}.cazikuf {color: var(--palette-brand-text);}.c1vqss5s {color: var(--palette-primary-text);}.cfr0jfq {color: var(--palette-secondary-text);}.c6a1giw {color: var(--palette-tertiary-text);}.c1u0bpqh {color: var(--palette-neutral-text);}.c1srkb0x {color: var(--palette-muted-text);}.cbrsnfr {color: var(--palette-danger-text);}.c1jf621i {color: var(--palette-warning-text);}.cz13af3 {color: var(--palette-negative-text);}.c58kddj {color: var(--palette-positive-text);}.cfg0erx {text-transform: capitalize;}.c1551sry {text-transform: lowercase;}.cpe440f {text-transform: uppercase;}.c1hyqsgu {font-weight: inherit;}.c1wm2u8a {font-weight: 100;}.c96w0js {font-weight: 300;}.czkiasv {font-weight: 400;}.c1b6fj3h {font-weight: 600;}.cveu3s3 {font-weight: 800;}.c1hg8m62 {font-family: var(--typography-font-monospace);}.c6u9ihh {font-family: var(--typography-font-heading);}.c9kwtas {line-height: var(--heading-l1-line-height);}.crgd4fe {font-size: var(--heading-l1-size);}.c1d7w45w {letter-spacing: var(--heading-l1-letter-spacing);}.c16vs593 {line-height: var(--heading-l2-line-height);}.c8hudd5 {font-size: var(--heading-l2-size);}.c15w79fb {letter-spacing: var(--heading-l2-letter-spacing);}.c1kouy52 {line-height: var(--heading-l3-line-height);}.c1nd9zqg {font-size: var(--heading-l3-size);}.c1qysrx2 {letter-spacing: var(--heading-l3-letter-spacing);}.c1v9whox {line-height: var(--heading-l4-line-height);}.c1rd0vjj {font-size: var(--heading-l4-size);}.c1y683ht {letter-spacing: var(--heading-l4-letter-spacing);}.c1vbmpn4 {line-height: var(--heading-l5-line-height);}.c1925epa {font-size: var(--heading-l5-size);}.ck7pk0g {letter-spacing: var(--heading-l5-letter-spacing);}.ctldzmb {line-height: var(--heading-l6-line-height);}.cq3mnn1 {font-size: var(--heading-l6-size);}.ccw0p9v {letter-spacing: var(--heading-l6-letter-spacing);}"`;
